In [1]:
from langchain_core.tools import tool

@tool
def add(a: int, b: int) -> int:
    """숫자 a, b를 더한다"""
    return a + b

@tool
def multiply(a: int, b: int) -> int:
    """숫자 a, b를 곱한다"""
    return a * b

In [8]:
add(2, 3)

TypeError: 'StructuredTool' object is not callable

In [2]:
add.invoke({'a': 2, 'b': 3})

5

In [3]:
from langchain_openai import ChatOpenAI

llm = ChatOpenAI(model='gpt-4o')

In [4]:
tools = [add, multiply]
llm_with_tools = llm.bind_tools(tools)

In [5]:
llm_with_tools.invoke('2 더하기 3은?')

AIMessage(content='', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 17, 'prompt_tokens': 83, 'total_tokens': 100, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_c7519a35cf', 'id': 'chatcmpl-DfJvmZuDVBds4uw5O8ji1jfUfG61O', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None}, id='lc_run--019e2531-1074-7263-9f7a-58187511f5d8-0', tool_calls=[{'name': 'add', 'args': {'a': 2, 'b': 3}, 'id': 'call_ISR5KCKCcF7k2AwGzVi1imM6', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 83, 'output_tokens': 17, 'total_tokens': 100, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

AIMessage(content='', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 17, 'prompt_tokens': 83, 'total_tokens': 100, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_5c172149cc', 'id': 'chatcmpl-Dd7fw8PHInpvzR1gXX6rYtCeADECK', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None}, id='lc_run--019e05f9-08a5-7123-b8bb-805ddad522d8-0', tool_calls=[{'name': 'add', 'args': {'a': 2, 'b': 3}, 'id': 'call_j9WbnXplp6GNQiGGCbAFe4xd', 'type': 'tool_call'}], invalid_tool_calls=[], usage_metadata={'input_tokens': 83, 'output_tokens': 17, 'total_tokens': 100, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

In [6]:
response = llm_with_tools.invoke('2 더하기 3은?')
print(response)
response.tool_calls

content='' additional_kwargs={'refusal': None} response_metadata={'token_usage': {'completion_tokens': 17, 'prompt_tokens': 83, 'total_tokens': 100, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_c7519a35cf', 'id': 'chatcmpl-DfJvqgWRh0sdtGGNjueVrxxYAifmW', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None} id='lc_run--019e2531-22ff-7ff3-acfc-69ebc1fbfbf6-0' tool_calls=[{'name': 'add', 'args': {'a': 2, 'b': 3}, 'id': 'call_g1o2rVGGPEt8uqyMDiFrv4mx', 'type': 'tool_call'}] invalid_tool_calls=[] usage_metadata={'input_tokens': 83, 'output_tokens': 17, 'total_tokens': 100, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}}


[{'name': 'add',
  'args': {'a': 2, 'b': 3},
  'id': 'call_g1o2rVGGPEt8uqyMDiFrv4mx',
  'type': 'tool_call'}]

In [7]:
from langchain_core.messages import AnyMessage, HumanMessage
from typing import Sequence

query = '2 더하기 3은?'

human_message = HumanMessage(query)
message_list: Sequence[AnyMessage] = [human_message]

In [8]:
ai_message = llm_with_tools.invoke(message_list)
print(ai_message)
ai_message.tool_calls

content='' additional_kwargs={'refusal': None} response_metadata={'token_usage': {'completion_tokens': 17, 'prompt_tokens': 83, 'total_tokens': 100, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_c7519a35cf', 'id': 'chatcmpl-DfJvy7VehSDPYSMtXSw6r1p3V5q6c', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None} id='lc_run--019e2531-4328-71a1-9e5f-531a196b086b-0' tool_calls=[{'name': 'add', 'args': {'a': 2, 'b': 3}, 'id': 'call_IVKqq6Hy10IRdL0DmROWBJWs', 'type': 'tool_call'}] invalid_tool_calls=[] usage_metadata={'input_tokens': 83, 'output_tokens': 17, 'total_tokens': 100, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}}


[{'name': 'add',
  'args': {'a': 2, 'b': 3},
  'id': 'call_IVKqq6Hy10IRdL0DmROWBJWs',
  'type': 'tool_call'}]

In [9]:
message_list.append(ai_message)

In [10]:
tool_message = add.invoke(ai_message.tool_calls[0])
tool_message

ToolMessage(content='5', name='add', tool_call_id='call_IVKqq6Hy10IRdL0DmROWBJWs')

In [11]:
message_list.append(tool_message)

In [12]:
llm_with_tools.invoke(message_list)

AIMessage(content='2 더하기 3은 5입니다.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 11, 'prompt_tokens': 108, 'total_tokens': 119, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_c7519a35cf', 'id': 'chatcmpl-DfJwBdpbFQIdTTKEXnQjMv2VQYtuZ', 'service_tier': 'default', 'finish_reason': 'stop', 'logprobs': None}, id='lc_run--019e2531-785c-75b3-942a-ee6270ed5257-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 108, 'output_tokens': 11, 'total_tokens': 119, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

AIMessage(content='2 더하기 3은 5입니다.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 11, 'prompt_tokens': 108, 'total_tokens': 119, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_5c172149cc', 'id': 'chatcmpl-Dd7tnbCyHz8nRtc2fKkhoj7tlhDf7', 'service_tier': 'default', 'finish_reason': 'stop', 'logprobs': None}, id='lc_run--019e0606-24a9-72d3-b78c-e3ee85aedebf-0', tool_calls=[], invalid_tool_calls=[], usage_metadata={'input_tokens': 108, 'output_tokens': 11, 'total_tokens': 119, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

In [13]:
tool_message_list = []

tools_by_name = {tool.name: tool for tool in tools}

for tool_call in ai_message.tool_calls:
    print(tool_call['name'])
    selected_tool = tools_by_name[tool_call['name']]
    tool_message = selected_tool.invoke(tool_call)
    tool_message_list.append(tool_message)

print(tool_message_list)

add
[ToolMessage(content='5', name='add', tool_call_id='call_IVKqq6Hy10IRdL0DmROWBJWs')]
